File: verify_test_user.py
#!/usr/bin/env python3
# filepath: /Users/daoming/Documents/Github/goAIME/verify_test_user.py
"""
Script to verify if test_user_good_performer exists and has sessions
"""

from pymongo import MongoClient
import sys

try:
    # Connect to MongoDB directly
    client = MongoClient('mongodb://localhost:27017')
    db = client.goaime
    
    # Check if the user exists
    user = db.users.find_one({"username": "test_user_good_performer"})
    if user:
        print(f"User test_user_good_performer exists with ID: {user['_id']}")
    else:
        print("User test_user_good_performer does not exist")
    
    # Check if the user has sessions
    session_count = db.sessions.count_documents({"username": "test_user_good_performer"})
    print(f"User has {session_count} sessions")
    
    # Check average accuracy
    if session_count > 0:
        pipeline = [
            {"$match": {"username": "test_user_good_performer"}},
            {"$group": {
                "_id": "$username",
                "avgAccuracy": {"$avg": "$accuracy"}
            }}
        ]
        
        avg_accuracy = list(db.sessions.aggregate(pipeline))
        if avg_accuracy:
            print(f"Average accuracy: {avg_accuracy[0].get('avgAccuracy', 0):.2f}%")
    
    # Print the first session for inspection
    if session_count > 0:
        print("\nFirst session sample:")
        session = db.sessions.find_one({"username": "test_user_good_performer"})
        
        # Print relevant information from the session
        print(f"  Session ID: {session['_id']}")
        print(f"  Contest: {session.get('contest', 'N/A')}")
        print(f"  Score: {session.get('score', 'N/A')}")
        print(f"  Accuracy: {session.get('accuracy', 'N/A')}")
        print(f"  Total correct: {session.get('total_correct', 'N/A')}")
        print(f"  Total attempted: {session.get('total_attempted', 'N/A')}")
        
        # Check how many problems are correct
        if 'problems_attempted' in session:
            problems = session['problems_attempted']
            correct_count = sum(1 for p in problems if p.get('is_correct', False))
            print(f"  Problems attempted: {len(problems)}")
            print(f"  Problems correct: {correct_count} ({(correct_count/len(problems)*100):.2f}%)")
    
except Exception as e:
    print(f"Error: {e}")
    sys.exit(1)

File: backend/scripts/verify_test_user.js
/**
 * Script to verify if test_user_good_performer exists in the database
 * Run with: node verify_test_user.js
 */

const mongoose = require('mongoose');
require('dotenv').config();

// Import models
const User = require('../models/User');
const Session = require('../models/Session');
const Problem = require('../models/Problem');

// Get MongoDB URI from environment variables
const MONGODB_URI = process.env.MONGODB_URI;

// Verify MongoDB URI
if (!MONGODB_URI) {
  console.error('MongoDB URI is not defined in .env file.');
  process.exit(1);
}

console.log('Connecting to MongoDB Atlas...');

// Connect to MongoDB
mongoose.connect(MONGODB_URI, {
  useNewUrlParser: true,
  useUnifiedTopology: true,
  serverSelectionTimeoutMS: 5000
})
.then(() => console.log('MongoDB Atlas connected successfully'))
.catch(err => {
  console.error('MongoDB connection error:', err);
  process.exit(1);
});

// Username to verify
const usernameToVerify = 'test_user_good_performer';

// Function to verify user existence and fetch a sample session
const verifyUser = async () => {
  try {
    // Check if user exists
    console.log(`Looking for user: ${usernameToVerify}`);
    const user = await User.findOne({ username: usernameToVerify });
    
    if (!user) {
      console.error(`User '${usernameToVerify}' not found in the database!`);
      console.log('Try running the create_test_user.js script first.');
      return;
    }
    
    console.log(`\n✅ USER FOUND:`);
    console.log(`ID: ${user._id}`);
    console.log(`Username: ${user.username}`);
    console.log(`Email: ${user.email}`);
    console.log(`Role: ${user.role}`);
    console.log(`Created: ${user.createdAt}`);
    
    // Count sessions for this user
    const sessionCount = await Session.countDocuments({ user: user._id });
    console.log(`\nSession count: ${sessionCount}`);
    
    if (sessionCount === 0) {
      console.log('No sessions found for this user.');
      return;
    }
    
    // Fetch a sample session
    const session = await Session.findOne({ user: user._id });
    
    console.log(`\n✅ SAMPLE SESSION:`);
    console.log(`ID: ${session._id}`);
    console.log(`Contest: ${session.contest} ${session.year}`);
    console.log(`Mode: ${session.mode}`);
    console.log(`Completed: ${session.completed}`);
    console.log(`Completed At: ${session.completedAt}`);
    console.log(`Score: ${session.score}`);
    console.log(`Total Attempted: ${session.totalAttempted}`);
    console.log(`Accuracy: ${session.accuracy}%`);
    console.log(`Average Time Per Problem: ${session.averageTimePerProblem}s`);
    
    // Count problems associated with this session
    const problemCount = await Problem.countDocuments({ session: session._id });
    console.log(`Problems in this session: ${problemCount}`);
    
    // Fetch a sample problem
    if (problemCount > 0) {
      const problem = await Problem.findOne({ session: session._id });
      
      console.log(`\n✅ SAMPLE PROBLEM:`);
      console.log(`ID: ${problem._id}`);
      console.log(`Number: ${problem.problemNumber}`);
      console.log(`Topic: ${problem.topic}`);
      console.log(`Difficulty: ${problem.difficulty}`);
      console.log(`Is Correct: ${problem.isCorrect}`);
      console.log(`Time Spent: ${problem.timeSpent}s`);
    }
    
    // Get topic performance summary
    console.log(`\n✅ TOPIC PERFORMANCE SUMMARY:`);
    const topics = await Problem.aggregate([
      { $match: { user: user._id } },
      { $group: {
          _id: "$topic",
          attempted: { $sum: 1 },
          correct: { $sum: { $cond: [{ $eq: ["$isCorrect", true] }, 1, 0] } }
        }
      },
      { $project: {
          topic: "$_id",
          attempted: 1,
          correct: 1,
          accuracy: { $multiply: [{ $divide: ["$correct", "$attempted"] }, 100] }
        }
      },
      { $sort: { attempted: -1 } }
    ]);
    
    topics.forEach(topic => {
      console.log(`${topic._id}: ${topic.correct}/${topic.attempted} correct (${topic.accuracy.toFixed(1)}%)`);
    });

  } catch (error) {
    console.error('Error verifying user:', error);
  } finally {
    mongoose.disconnect();
    console.log('\nMongoDB disconnected');
  }
};

// Run the verification
verifyUser();
